File: src/main/java/com/mynimef/bot/config/CommandsBuilder.java
package com.mynimef.bot.config;

import com.mynimef.bot.IBot;
import com.mynimef.bot.actions.IAction;

import java.util.HashMap;
import java.util.Map;

public abstract class CommandsBuilder {

    private final Map<String, IAction> commands;

    public CommandsBuilder() {
        this.commands = new HashMap<>();
    }

    protected abstract void init();

    protected void add(String command, String reply) {
        commands.put(command, (
                String text,
                String chatId,
                Integer messageId,
                String username,
                String firstName,
                String lastName,
                IBot bot
        ) -> bot.sendMessage(chatId, reply));
    }

    protected void add(String command, IAction action) { commands.put(command, action); }

    public Map<String, IAction> getCommands() {
        init();
        return commands;
    }

    protected abstract void nonCommandUpdate(
            String text,
            String chatId,
            Integer messageId,
            String username,
            String firstName,
            String lastName,
            IBot bot
    );

    public IAction getNoCommandRecognized() {
        return this::nonCommandUpdate;
    }
}

File: src/main/kotlin/com/mynimef/bot/config/CommandsBuilder.kt
package com.mynimef.bot.config

import com.mynimef.bot.IBot
import com.mynimef.bot.containers.UserCommand
import com.mynimef.bot.executable.Action


abstract class CommandsBuilder {

    private val _commands: MutableMap<String, Action> = HashMap()
    internal val commands: Map<String, (command: UserCommand, bot: IBot) -> Unit> by lazy {
        init()
        _commands
    }

    protected abstract fun init()

    protected fun add(command: String, reply: String) {
        _commands[command] = { userCommand, bot ->
            bot.sendMessage(userCommand.chatId, reply)
        }
    }

    protected fun add(command: String, action: Action) {
        _commands[command] = action
    }

    abstract val nonCommandUpdate: Action

}